rfc/rfc-66/rfc-66.md

```diff
@@ -0,0 +1,239 @@
+| Reader expectations             | Readers still query based on file slices, see the pseudocode for new file slicing                                                             |
+| Synchronization and correctness | Still need some mutex on commit log (timeline). Getting the completion time and adding to the timeline needs to happen atomically             |
+
+**Note:** Clustering and updates conflict resolution is not in the scope of non-blocking concurrency control, you still need the OCC for course-grained failures and retries. 
+
+## Appendix
+
+### Pseudocode for New File Slicing
+
+```python
+# new file slicing
+file_slice_barriers = []
+# sort the base file by start time reversely and get the barriers
+for base_file in fg.getBaseFilesSortedReverselyByStartTime():
+  file_slice_barriers.add(instant_time(base_file))
+
+file_slices = []
+for log_file in fg.getLogFilesSortedByCompletionTime():
+  completion_time = completion_time(log_file)
+  file_slice = find_file_slice(completion_time, file_slice_barriers)
+  file_slices.add(file_slice)
+
+def find_file_slice(completion_time, file_slice_barriers):
+  # Note: needs to handle special case where file_slice_barriers is empty (no base file)
+  for barrier in file_slice_barriers:
+    if (barrier < completion_time):
+      # each file slice is attached to a barrier, returns the existing file slice or a fresh new one based on the barrier.
+      return get_or_create_file_slice(barrier)
+```
+
+#### A Demo Conducive to Comprehension
+
+Just to make it concrete for the readers, we can add an example.
+Let's take this example of one filegroup and consider the following sequence of commits:
+
+```xml
+fg_t10 -> very first commit with start time, end time as [t10, t20].
+l1 -> log file version 1 start time, end time as [t21, t40].
+l2 -> concurrent log file version 2 [t30, t50].
+fg_t60 -> base file due to compaction [t60, t80].
+l3 -> concurrent log file version 3 [t35, t90].
+```
+
+In this case, file_slice_barriers list is [t60, t10]. For a query at t100, getAllFileSlices should return the following list:
+
+```xml
+[
+{t60, fg_t60.parquet, {l3}},
+{t10, fg_t10.parquet, {l1, l2}}
+]
+```
+
+This assumes that file slicing is done based on completion time.
+
+### Pseudocode for Time Travel Queries
+
+```python
+# This is based on the new file slicing.
+# find the write 'w' with completionTime(w) < t, among all the writes that completed before t 
+# (Note t may not match an existing completion time directly) 
+max_completion_time = findMaxCompletedTimeBefore(t) 
+
+fileSlicesToRead = [] 
+for fg in fileGroups:
+  fileSlice = latestFileSliceThatIsOverlappingWithCompletionTime(fg, max_completion_time)
+  fileSlicesToRead.append(fileSlice)
+return fileSlicesToRead
+
+def latestFileSliceThatIsOverlappingWithCompletionTime(fg, max_completion_time):
+  # getAllFileSlices() returns the file slices as it does today ordered reversely by instant time.
+  for f_slice in fg.getAllFileSlices():
```

Comment:
Are these guaranteed to be non-overlapping ie a given log file or base file occurs in only one slice?